addons/sale_crm/models/sale_order.py

```diff
@@ -13,3 +13,23 @@ class SaleOrder(models.Model):
 
     def action_confirm(self):
         return super(SaleOrder, self.with_context({k:v for k,v in self._context.items() if k != 'default_tag_ids'})).action_confirm()
+
+    def _message_post_after_hook(self, new_message, message_values):
+        self._set_expected_revenue()
+        return super()._message_post_after_hook(new_message, message_values)
+
+    def action_cancel(self):
+        res = super().action_cancel()
+        self._set_expected_revenue()
```

Comment:
Since the logic we add is actually modifying the lead, I would rather add a method on the crm.lead model and call it where appropriate with

`self.opportunity_id._update_expected_revenues_from_so()`

and then in the method, start with

for opportunity in self:
... UPDATE CODE ....

so that if the opp. has no lead, it still work properly. This would become the only method overriden in crm_sale_subscription.